src/dlsproto/node/neo/internal/SuspendableRequest.d

```diff
@@ -504,33 +506,33 @@ public struct SuspendableRequest
         ReceivedMessageAction msg_action;
         do
         {
-            int resume_code;
+            RequestOnConnBase.EventDispatcher.EventNotification event;
 
             this.fiber_suspended_waiting_for_data = true;
             try
             {
-                resume_code = this.conn.receiveAndHandleEvents(
-                    ( in void[] received )
-                    {
-                        msg_action = handle_received_message(received);
-                    }
-                );
+                event =  this.conn.nextEvent(this.conn.NextEventFlags.Receive);
+
+                if (event.active == event.active.received)
+                {
+                    msg_action = handle_received_message(event.received.payload);
+                }
             }
             finally
             {
                 this.fiber_suspended_waiting_for_data = false;
             }
 
             // Fiber was resumed with the data ready code.
-            if ( resume_code > 0 ) // positive code => user code => must be data ready
+            if ( event.active == event.active.resumed ) // positive code => user code => must be data ready
```

Comment:
Not sure about the `if` here. What other cases are you trying to cover?

Reply:
So, we're either resumed from the client (line 535), or we're resumed by the storage engine (by calling fiber.resume with the `data_ready_code`). Does this make sense?

Reply:
Ah yes I see. A `switch` would make more sense now, I think.

Reply:
`switch`?